### PR TITLE
chore(ci): rework build-tracker to use redis instead of in-memory store of build results

### PR DESCRIPTION
**Pull request by Strum355**
Currently, build-tracker keeps track of consecutive build failures through an in-memory store of failed builds. As this gets deployed more frequently on MSP, we lose state more frequently which would result in incorrect results. Instead, we can use redis as our external store as well as for locking using redsync

## Test plan

Unit tests have been updated, but proper testing will require live traffic

### CODE REVIEW

File: dev/build-tracker/build/build.go [hunk 1]
```diff
@@ -226,45 +228,85 @@ func (b *Event) GetBuildNumber() int {
...
-		builds:              make(map[int]*Build),
-		consecutiveFailures: make(map[string]int),
+		r:  rclient,
+		m1: lock,
+	}
+}
 
-		m: sync.RWMutex{},
+func (s *Store) lock() (func(), error) {
+	for {
```
Comment (Strum355) [inline dev/build-tracker/build/build.go:255]: probably want to bound this looping

File: dev/build-tracker/build/build.go [hunk 2]
```diff
@@ -226,45 +226,79 @@ func (b *Event) GetBuildNumber() int {
...
-		s.builds[event.GetBuildNumber()] = build
 	}
 
-	// Now that we have a build, lets make sure it isn't modified while we look and possibly update it
-	build.Lock()
-	defer build.Unlock()
+	// write out the build to redis at the end, once all mutations are applied
+	defer func() {
+		buildb, _ = json.Marshal(build)
+		s.r.Set(context.Background(), "build/"+strconv.Itoa(event.GetBuildNumber()), buildb, 0)
```
Comment (jhchabran) [inline dev/build-tracker/build/build.go:294]: We're repeating `context.Background()` many times in the code, why don't we assign it at the top at least? Ideally though, we should get the context from app itself to handle graceful shutdown.

File: dev/build-tracker/main.go [hunk 3]
```diff
@@ -423,7 +422,14 @@ func (s Service) Initialize(ctx context.Context, logger log.Logger, contract run
 		return nil, err
 	}
 
-	server := NewServer(fmt.Sprintf(":%d", contract.Port), logger, config, bqWriter)
+	redisOpts, err := redis.ParseURL(*contract.RedisEndpoint)
+	if err != nil {
+		return nil, err
+	}
+	rclient := redis.NewClient(redisOpts)
+	rlock := redsync.New(goredis.NewPool(rclient)).NewMutex("build-tracker", redsync.WithExpiry(time.Second*30))
```
Comment (jhchabran) [inline dev/build-tracker/main.go:430]: Right, I was about to ask how do we ensure we don't deadlock ourselves if something goes wrong, but the 30s timeout is 👍

**Review by jhchabran — APPROVED** [review]
LGTM. Could you add in the README a link to the msp docs to see how to deal with Redis in case something goes wrong? And could you mention along the way that it'd fine to nuke it if hypothetically something went wrong?

File: dev/build-tracker/build/build.go (hunk 2 above)
Comment (Strum355) [reply dev/build-tracker/build/build.go]: Have introduced context into the handlers, this should help better with the 30s expiry on the redsync locks as well :+1:

**Pull request merged**